pong.h

```diff
@@ -12,12 +12,36 @@ namespace pong
 class Pong
 {
 private:
-    std::vector<std::unique_ptr<GameObject>>mGameObjects {};
+    Pong() = default;
+    Pong(const Pong&) = delete;
+    Pong& operator=(const Pong&) = delete;
+    Pong(Pong&&) = delete;
+    Pong& operator=(Pong&&) = delete;
+    ~Pong() = default;
+
+    std::vector<std::shared_ptr<GameObject>>mGameObjects {};
     CollisionManager mCollisionManager {};
 
 public:
-    void PongInit();
-    void PongGameLoop();
+    static Pong& GetInstance();
+
+    static void Init();
+    static void GameLoop();
+    static void Cleanup();
+
+    template<typename T>
+    static std::shared_ptr<T> FindGameObject()
```

Comment:
This is going to be relatively costly at runtime to do, dynamic casts are very expensive.

The [google coding guidelines](https://google.github.io/styleguide/cppguide.html) advise against dynamic_cast and "Run Time Type Inference" (RTTI): 

RTTI has legitimate uses but is prone to abuse, so you must be careful when using it. You may use it freely in unittests, but avoid it when possible in other code. In particular, think twice before using RTTI in new code. If you find yourself needing to write code that behaves differently based on the class of an object, consider one of the following alternatives to querying the type:

    Virtual methods are the preferred way of executing different code paths depending on a specific subclass type. This puts the work within the object itself.
    If the work belongs outside the object and instead in some processing code, consider a double-dispatch solution, such as the Visitor design pattern. This allows a facility outside the object itself to determine the type of class using the built-in type system.

Reply:
Well in that case what is a better way to do this? This function is mainly to find specific gameobjects during runtime. Like with the Opponent it needs a reference to the ball gameobject so finds it with this function in OnStart(). There certainly could be something like like `FindGameObject("Ball")` keying on its name, but I feel that has its own problems. 

Basing things off of Unity (suprise, suprise) this is a similar interface
```
public void Start()
{
     mPlayer = FindObjectOfType<Player>()

     // Or for many objects
     List<Enemy> enemies = FindObjectsOfType<Enemy>();
}
```

Reply:
interesting. I would say spend a second looking into unity and seeing if you can determine what they are doing under the hood. If there is prior art for this, let 'r' rip :)

Reply:
Yeah, it is hard to say exactly how it works in Unity since it is private. But they do make the disclaimer
`Note: This function is very slow. It is not recommended to use this function every frame. In most cases you can use the singleton pattern instead.` So I do imagine it could be something similar?

I did take a look at Unreal Engine as well since their code is open. I couldn't quite understand entirely how it was working. I think they are doing something much more clever to determine types but they make a similar disclaimer.
```
/** 
*	Find all Actors in the world of the specified class. 
*	This is a slow operation, use with caution e.g. do not use every frame.
*/
static ENGINE_API void GetAllActorsOfClass(const UObject* WorldContextObject, TSubclassOf<AActor> ActorClass, TArray<AActor*>& OutActors);
```
I have some ideas that I will look into to do this better based off of what I gathered from Unreal Engine, but that will take some time to flesh out.

Reply:
Yeah interesting. I think these likely both indicate a dynamic cast under the hood.

Definitely no rush on figuring any of it out, just a good thing to be aware of the trade offs :)